### PR TITLE
Use specific chatters for signalling

### PR DESCRIPTION
**Pull request by zippy**
Now send signals only to recent chatters in the a channel's history rather than to all active agents, except in the case that there isn't any history.

### CODE REVIEW

File: src/applications/ElementalChat/store/elementalChat.js [hunk 1]
```diff
@@ -103,6 +104,36 @@ function _addMessageToChannel(rootState, commit, state, channel, message) {
...
+    const msg = channel.messages[i];
+    const count = Object.keys(chatters).length;
+    // absolute count limit on who to send signals to
+    if (count > MAX_CHATTERS_FOR_SIGNAL) {
+      break;
+    }
+    // when to stop going backward in the history looking for people too send to
+    if (
+      now - new Date(msg.createdAt[0] * 1000) > MAX_MESSAGE_AGE_FOR_SIGNAL &&
+      count > MIN_CHATTERS_FOR_SIGNAL
```
Comment (timotree3) [inline src/applications/ElementalChat/store/elementalChat.js:124]:
```suggestion
      count >= MIN_CHATTERS_FOR_SIGNAL
```

File: src/applications/ElementalChat/store/elementalChat.js [hunk 2]
```diff
@@ -103,6 +104,36 @@ function _addMessageToChannel(rootState, commit, state, channel, message) {
...
+    ) {
+      break;
+    }
+    const agentStr = arrayBufferToBase64(msg.createdBy);
+    if (agentStr != meStr) {
+      chatters[agentStr] = msg.createdBy;
+    }
+  }
+  console.log("chatters for signals:", Object.values(chatters));
```
Comment (timotree3) [inline src/applications/ElementalChat/store/elementalChat.js:133]:
```suggestion
  console.log("chatters to be signaled:", Object.values(chatters));
```

**Review by timotree3 — APPROVED** [review]
I approve minus two small changes and assuming that this is the approach we want to take

File: src/applications/ElementalChat/store/elementalChat.js [hunk 3]
```diff
@@ -103,6 +104,36 @@ function _addMessageToChannel(rootState, commit, state, channel, message) {
...
+  let chatters = {};
+  const now = Date.now();
+  const meStr = arrayBufferToBase64(me);
+  for (let i = channel.messages.length - 1; i > 0; i--) {
+    const msg = channel.messages[i];
+    const count = Object.keys(chatters).length;
+    // absolute count limit on who to send signals to
+    if (count > MAX_CHATTERS_FOR_SIGNAL) {
+      break;
+    }
```
Comment (robbiecarlton) [inline src/applications/ElementalChat/store/elementalChat.js:120]: NBD but you could also put this in the condition of the for loop

File: src/applications/ElementalChat/store/elementalChat.js [hunk 4]
```diff
@@ -103,6 +104,36 @@ function _addMessageToChannel(rootState, commit, state, channel, message) {
...
+    // when to stop going backward in the history looking for people too send to
+    if (
+      now - new Date(msg.createdAt[0] * 1000) > MAX_MESSAGE_AGE_FOR_SIGNAL &&
+      count >= MIN_CHATTERS_FOR_SIGNAL
+    ) {
```
Comment (robbiecarlton) [inline src/applications/ElementalChat/store/elementalChat.js:125]: same, although at that point it starts getting pretty unwieldy ;)

File: src/applications/ElementalChat/store/elementalChat.js [hunk 5]
```diff
@@ -346,7 +377,18 @@ export default {
             messageData: message,
             channelData: payload.channel
           };
-          dispatch("signalMessageSent", signalMessageData);
+          const chatters = getRecentChatters(
+            rootState.agentKey,
+            payload.channel
+          );
+          if (chatters.length == 0) {
+            dispatch("signalMessageSent", signalMessageData);
```
Comment (robbiecarlton) [inline src/applications/ElementalChat/store/elementalChat.js:385]: so if there are no recent chatters we broadcast to everyone?

**Review by robbiecarlton — APPROVED** [review]
One question about functionality. Code looks good

**Comment by zippy** [comment]
incorporated elsewhere

**Pull request closed without merging**